compiler/rustc_lint/src/default_could_be_derived.rs

```diff
@@ -0,0 +1,422 @@
+            let hir::ExprKind::Block(hir::Block { stmts: _, expr: Some(expr), .. }, None) =
+                body.value.kind
+            else {
+                continue;
+            };
+
+            // We check `fn default()` body is a single ADT literal and all the fields are being
+            // set to something equivalent to the corresponding types' `Default::default()`.
+            match expr.kind {
+                hir::ExprKind::Struct(_qpath, fields, tail) => {
+                    // We have a struct literal
+                    //
+                    // struct Foo {
+                    //     field: Type,
+                    // }
+                    //
+                    // impl Default for Foo {
+                    //     fn default() -> Foo {
+                    //         Foo {
+                    //             field: val,
+                    //         }
+                    //     }
+                    // }
+                    //
+                    // We suggest #[derive(Default)] if
+                    //  - `field` has a default value, regardless of what it is; we don't want to
+                    //    encourage divergent behavior between `Default::default()` and `..`
+                    //  - `val` is `Default::default()`
+                    //  - `val` matches the `Default::default()` body for that type
+                    //  - `val` is `0`
+                    //  - `val` is `false`
+                    if let hir::StructTailExpr::Base(_) = tail {
+                        // This is *very* niche. We'd only get here if someone wrote
+                        // impl Default for Ty {
+                        //     fn default() -> Ty {
+                        //         Ty { ..something() }
+                        //     }
+                        // }
+                        // where `something()` would have to be a call or path.
+                        return;
+                    }
+                    // All the fields that have been provided can be used by `#[derive(Default)]`
+                    // FIXME: check if a field is suitable to be given to the default field value.
+                    let all_defaultable = fields.iter().all(|f| {
+                        orig_fields.get(&f.ident.name).and_then(|f| f.default).is_some()
+                            || check_expr(cx, f.expr)
+                    });
+                    // The type has default field values.
+                    #[allow(rustc::potential_query_instability)]
+                    let any_default_field = orig_fields.iter().any(|(_, f)| f.default.is_some());
+                    // The `Default::default()` overrides any of the default field values.
+                    let any_default_field_given = fields
+                        .iter()
+                        .any(|f| orig_fields.get(&f.ident.name).and_then(|f| f.default).is_some());
+
+                    if all_defaultable && any_default_field {
```

Comment:
The check that the underlying type has default fields. The else check also checks that.